### ISSUE TITLE
Prometheus WAL Size Keeps Growing Indefinitely 

### ISSUE THREAD

**Issue by wazery** [prometheus/prometheus#7955]
**What did you do?**

Currently, I have a Prometheus v2.20.0 running, and it has an issue that the WAL keeps on growing indefinitely and consuming disk space.

Actually the disk space is not the issue now, but that the WAL folder is not getting cleaned, so if any time Prometheus is restarted, it tries to load the entire WAL into memory.

So for example WAL is now 60GB, and memory is 32GB, so Prometheus keeps on restarting when it gets killed by the OOM killer, as it consumes the whole server memory of 32GB.

Here is my current config for it, and please note that I run it using Docker Compose.

       - '--web.enable-admin-api'
       - '--config.file=/etc/prometheus/prometheus.yml'
       - '--web.external-url=https://prometheus.example.com'
       - '--storage.tsdb.path=/var/lib/prometheus'
       - '--storage.tsdb.retention=150d'
       - '--web.console.libraries=/usr/share/prometheus/console_libraries'
       - '--web.console.templates=/usr/share/prometheus/consoles'

**What did you expect to see?**

I expect by the above configurations and the current setup, the WAL folder got cleared by itself and stops on growing indefinitely.

**What did you see instead? Under which circumstances?**

WAL folder keeps on growing indefinitely till consuming the entire server disk. 

**Environment**

* System information:

`Linux 4.19.0-10-cloud-amd64 x86_64`

* Prometheus version:
```
prometheus, version 2.20.0 (branch: HEAD, revision: e5a06b483527d4fe0704b8fa3a2b475b661c526f)
  build user:       root@ac954b6d5c6e
  build date:       20200722-18:51:45
  go version:       go1.14.6
```
* Alertmanager version:
```
alertmanager, version 0.18.0 (branch: HEAD, revision: 1ace0f76b7101cccc149d7298022df36039858ca)
  build user:       root@868685ed3ed0
  build date:       20190708-14:31:49
  go version:       go1.12.6
```
* Prometheus configuration file:
```
       - '--web.enable-admin-api'
       - '--config.file=/etc/prometheus/prometheus.yml'
       - '--web.external-url=https://prometheus.<truncated>'
       - '--storage.tsdb.path=/var/lib/prometheus'
       - '--storage.tsdb.retention=150d'
       - '--web.console.libraries=/usr/share/prometheus/console_libraries'
       - '--web.console.templates=/usr/share/prometheus/consoles'
```

* Alertmanager configuration file:
```
      - '--config.file=/etc/alertmanager/alertmanager.yml'
      - '--web.external-url=https://alertmanager.<truncated>'
      - '--storage.path=/alertmanager'
      - '--log.level=debug'
```

**Comment by codesome** [comment]
@wazery can you share the logs of Prometheus?

**Comment by wazery** [comment]
@codesome I've checked the logs and it's all internal logs, nothing is related to WAL or similar manners.

A snippet from it (which is repeated for similar issues and nothing else is there)
```
level=error ts=2020-09-21T09:35:00.384Z caller=consul.go:487 component="discovery manager scrape" discovery=consul msg="Error refreshing service" service=jm tags= err="Get \"http://consul.cloud:8500/v1/health/service/jm?index=17326658&wait=600000ms\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)"
level=error ts=2020-09-21T09:35:00.385Z caller=consul.go:487 component="discovery manager scrape" discovery=consul msg="Error refreshing service" service=docker-engine-tags= err="Get \"http://consul.cloud:8500/v1/health/service/docker-engine?index=19110756&wait=600000ms\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)"
```

**Comment by codesome** [comment]
@wazery are there any error logs related to compaction or checkpointing? (or any warning/error)

**Comment by wazery** [comment]
@codesome Checked now, and it seems yes, there are some logs related to compaction:

```
level=info ts=2020-09-21T09:00:21.916Z caller=compact.go:495 component=tsdb msg="write block" mint=1600668000000 maxt=1600675200000 ulid=01EJQY8APBT3WDSKXPW6128GTT duration=21.840937418s
level=info ts=2020-09-21T09:00:24.856Z caller=head.go:804 component=tsdb msg="Head GC completed" duration=1.996847498s
level=error ts=2020-09-21T09:00:24.857Z caller=db.go:685 component=tsdb msg="compaction failed" err="reload blocks: head truncate failed: truncate chunks.HeadReadWriter: maxt of the files are not set"
```

**Comment by wazery** [comment]
Also checkointing:

```
level=info ts=2020-09-18T15:00:02.077Z caller=compact.go:495 component=tsdb msg="write block" mint=1600430400000 maxt=1600437600000 ulid=01EJGVNCFXGKKNDZ3XQ7ZX3D71 duration=928.688194ms
level=info ts=2020-09-18T15:00:02.236Z caller=head.go:804 component=tsdb msg="Head GC completed" duration=32.686222ms
level=info ts=2020-09-18T15:00:02.268Z caller=checkpoint.go:96 component=tsdb msg="Creating checkpoint" from_segment=3951 to_segment=3953 mint=1600437600000
level=info ts=2020-09-18T15:00:02.702Z caller=head.go:884 component=tsdb msg="WAL checkpoint complete" first=3951 last=3953 duration=433.990534ms
```

**Comment by roidelapluie** [comment]
Can you upgrade to 2.21? Thanks!

**Comment by wazery** [comment]
@roidelapluie So it's a known bug and the upgrade will solve it?

**Comment by codesome** [comment]
@wazery yes a known bug which is fixed in v2.21

**Comment by wazery** [comment]
@codesome and @roidelapluie many many thanks for your support. I will upgrade and report here then will close the issue.

Wish you both an awesome day!

**Comment by roidelapluie** [comment]
Here is the fix: #7856

**Comment by beorn7** [comment]
Assuming that this is indeed fixed now, I'll close it. Please reopen if you continue to see this issue. Thank you!

**Comment by beorn7** [comment]
u

**Comment by tomjohnburton** [comment]
We are running `v2.23.0` and this problems seems to still be present.

![image](https://user-images.githubusercontent.com/39559415/101380355-abf30a80-38b5-11eb-9cf9-949309579f57.png)

**Comment by codesome** [comment]
@tomjohnburton this issue is regarding WAL growth and  no the memory. If in your case it was memory growth without increase in number of series, can you open another issue with more details?

**Comment by Tafsiralam** [comment]
From above comment 

> @wazery yes a known bug which is fixed in v2.21

Should I upgrade to the latest stable version v2.26 to fix the WAL size issue? or should I go for v2.21?
